src/decoders/hexadecimal_decoder.rs

```diff
@@ -0,0 +1,187 @@
+use crate::checkers::CheckerTypes;
+use crate::decoders::interface::check_string_success;
+
+use super::crack_results::CrackResult;
+use super::interface::Crack;
+use super::interface::Decoder;
+
+use log::{info, trace};
+
+///! Hexadecimal Decoder
+pub struct HexadecimalDecoder;
+
+impl Crack for Decoder<HexadecimalDecoder> {
+    fn new() -> Decoder<HexadecimalDecoder> {
+        Decoder {
+            name: "hexadecimal",
+            description: "Data is broken into 4-bit sequences, and each value (between 0 and 15 inclusively) is encoded using one of 16 symbols from the ASCII character set. Although any 16 symbols from the ASCII character set can be used, in practice the ASCII digits '0'–'9' and the letters 'A'–'F' (or the lowercase 'a'–'f') are always chosen in order to align with standard written notation for hexadecimal numbers.",
+            link: "https://en.wikipedia.org/wiki/Hexadecimal#Base16_(transfer_encoding)",
+            tags: vec!["hexadecimal", "hex", "base", "decoder"],
+            expected_runtime: 0.01,
+            expected_success: 1.0,
+            failure_runtime: 0.01,
+            normalised_entropy: vec![1.0, 10.0],
+            popularity: 1.0,
+            phantom: std::marker::PhantomData,
+        }
+    }
+
+    /// This function does the actual decoding
+    /// It returns an Option<string> if it was successful
+    /// Else the Option returns nothing and the error is logged in Trace
+    fn crack(&self, text: &str, checker: &CheckerTypes) -> CrackResult {
+        trace!("Trying hexadecimal with text {:?}", text);
+        let decoded_text = hexadecimal_to_string(text);
+        let mut results = CrackResult::new(self, text.to_string());
+
+        trace!("Decoded text for hexadecimal: {:?}", decoded_text);
+
+        if !check_string_success(&decoded_text, text) {
+            info!(
+                "Failed to decode hexadecimal because check_string_success returned false on string {}",
+                decoded_text
+            );
+            return results;
+        }
+
+        let checker_result = checker.check(&decoded_text);
+        results.unencrypted_text = Some(vec![decoded_text]);
+
+        results.update_checker(&checker_result);
+
+        results
+    }
+}
+
+/// Decodes hexadecimal to string
+fn hexadecimal_to_string(hex: &str) -> String {
+    // Remove all non-hexadecimal characters from the string
+    let hex = hex.replace(|c: char| !c.is_ascii_hexdigit(), "");
```

Comment:
If a string has non hex characters, does that mean it's not hex and we should skip it?